imap_processing/codice/codice_l2.py

```diff
@@ -866,6 +957,162 @@ def process_hi_sectored(dependencies: ProcessingInputCollection) -> xr.Dataset:
+    energy_kev = np.full(apd_energy.shape, np.nan)
+    # The rows are apd_energy bins
+    energy_bins_inds[valid_mask] = energy_table[
+        apd_energy[valid_mask], col_inds[valid_mask]
+    ]
+    energy_kev[valid_mask] = energy_bins[energy_bins_inds[valid_mask].astype(int)]
+
+    l2_dataset["apd_energy"].data = (
+        np.array(energy_kev).astype(np.float32).reshape(apd_energy_shape)
+    )
+
+    # Calculate TOF in nanoseconds
+    tof_bit_to_ns = get_mpq_calc_tof_conversion_vals(dependencies)
+    tof_bits = l2_dataset["tof"].values.flatten()
+    # Create output array
+    tof_ns = np.full(tof_bits.shape, np.nan, dtype=np.float64)
+    # Get only valid TOF bits between 0 and 1023
+    valid_mask = (tof_bits >= 0) & (tof_bits < 1024)
+    tof_ns[valid_mask] = tof_bit_to_ns[tof_bits[valid_mask]]
+    # Reshape back to original shape
+    l2_dataset["tof"].data = tof_ns.astype(np.float32).reshape(l2_dataset["tof"].shape)
+
+    # Convert energy step to energy in keV
+    esa_kev = get_mpq_calc_energy_conversion_vals(dependencies)
+    energy_steps = l2_dataset["energy_step"].values.flatten()
+    # Create output array
+    kev = np.full(energy_steps.shape, np.nan, dtype=np.float64)
+    # Get only valid energy_steps between 0 and 128
+    valid_mask = (energy_steps >= 0) & (energy_steps < 128)
+    kev[valid_mask] = esa_kev[energy_steps[valid_mask]]
+    # Reshape back to original shape
+    l2_dataset["energy_per_charge"] = (
+        l2_dataset["energy_step"].dims,
+        kev.astype(np.float32).reshape(l2_dataset["energy_step"].shape),
+    )
+    # Drop unused variables
+    vars_to_drop = ["spare", "sw_bias_gain_mode", "st_bias_gain_mode", "k_factor"]
+    l2_dataset = l2_dataset.drop_vars(vars_to_drop)
+    # Update variable attributes
+    l2_dataset.attrs.update(
+        cdf_attrs.get_global_attributes("imap_codice_l2_lo-direct-events")
+    )
+    for var in l2_dataset.data_vars:
+        l2_dataset[var].attrs.update(cdf_attrs.get_variable_attributes(var))
+    # Update coord attributes
+    l2_dataset["priority"].attrs.update(
+        cdf_attrs.get_variable_attributes("priority", check_schema=False)
+    )
+    l2_dataset["event_num"].attrs.update(
+        cdf_attrs.get_variable_attributes("event_num", check_schema=False)
+    )
+    l2_dataset["epoch"] = xr.DataArray(
+        l2_dataset["epoch"].data,
+        dims="epoch",
+        attrs=cdf_attrs.get_variable_attributes("epoch", check_schema=False),
+    )
+    l2_dataset["epoch_delta_minus"] = xr.DataArray(
+        data=l2_dataset["epoch_delta_minus"].data.astype(np.int64),
+        dims="epoch",
+        attrs=cdf_attrs.get_variable_attributes(
+            "epoch_delta_minus", check_schema=False
+        ),
+    )
+    l2_dataset["epoch_delta_plus"] = xr.DataArray(
+        l2_dataset["epoch_delta_plus"].data.astype(np.int64),
+        dims="epoch",
+        attrs=cdf_attrs.get_variable_attributes("epoch_delta_plus", check_schema=False),
+    )
```

Comment:
For these, should we update the l1a data to be "correct" so you don't have to mess with anything here? I would have expected these to just pass through if you aren't doing anything to the values, but you are changing the datatype and changing the attributes.